### PR TITLE
[Draft] 2909 movement of structure ceases when it reaches the boundaries of canvas requiring the consistent movement of mouse to continue its motion

### LINKED ISSUE #2909: Movement of structure ceases when it reaches the boundaries of canvas, requiring the consistent movement of mouse to continue its motion
**Steps to Reproduce**
1. Open Ketcher
2. Add two Benzene rings on canvas
3. Select one of them and move to the edges of canvas

**Actual behavior**
The movement of the structure occurs in shaking when the user brings the structure close to the edge of the canvas, and in order for the structure to continuously move beyond the canvas boundaries, continuous mouse movements are required.

**Expected behavior**
When the user brings the structure close to the edge of the canvas, smooth and continuous movement should be ensured instead of shaking motion. Additionally, the structure should be able to move beyond the canvas boundaries without the need for continuous mouse movements. 

**Screenshots**

https://github.com/epam/ketcher/assets/62464241/289f3432-1baa-4293-a9aa-9222fbfe79e6

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser Chrome
 - Version 112.0.5615.138 (Official Build) (64-bit)

**Ketcher version** 
[Version 2.12.0-rc.3](https://lifescience.opensource.epam.com/ketcher/index.html)

### PR DESCRIPTION
**Pull request by NatalliaHutovich**
Resolves #2909 
Allow run canvas automatic scroll and resizing. Selected area is moving and set near the edge of canvas or screen, and mouse crosses visible area edge and stay pressed  on the top/right/bottom/left panels. If selected area is near the canvas edge, size of canvas is increased and visible area is scrolled to new edge.  If selected area is near the edge of visible area (screen), visible area is scrolled. Automatic scroll is finished when mouse is released or leaves the document.

## Check list
- [ ] unit-tests written
- [ ] e2e-tests written
- [ ] documentation updated
- [ ] PR name follows the pattern `#1234 – issue name`
- [x] branch name doesn't contain '#'
- [x] PR is linked with the issue
- [x] base branch (master or release/xx) is correct
- [x] task status changed to "Code review"

### CODE REVIEW

File: packages/ketcher-core/src/application/render/raphaelRender.ts [hunk 1]
```diff
@@ -214,7 +213,7 @@ export class Render {
 
       const isAutoScale = this.options.autoScale || this.options.downScale;
       if (!isAutoScale) {
-        const ext = Vec2.UNIT.scaled(scale);
+        const ext = Vec2.UNIT.scaled(12);
```
Comment (NatalliaHutovich) [inline packages/ketcher-core/src/application/render/raphaelRender.ts:216]: ToDo: This value influences on distance between structure and canvas edge and should be made conditional.  If selection is moved, should have a small value (like 5 ) in order to structure always be near the edge of canvas. In other cases, should have 'scale' value = 40 as it was before, in order to structure has enough space before canvas edge for selection lasso, rotate controller and so on.

File: packages/ketcher-react/src/script/editor/utils/canvasExtension.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    if (vector.y < 0) {
+      clientArea.scrollTop = 0;
+    }
+    if (vector.y > 0) {
+      clientArea.scrollTop = clientArea.scrollHeight;
+    }
+  });
+}
+
+export function scrollByDirection(editor: Editor, direction) {
```
Comment (NatalliaHutovich) [inline packages/ketcher-react/src/script/editor/utils/canvasExtension.ts:134]: When structure is near the edge of screen (not canvas) and is moved out of visible area, clientArea is scrolled to follow the structure. You can influence on scroll speed using scrollMultiplier

File: packages/ketcher-react/src/script/editor/utils/canvasExtension.ts [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+  const { leftX, topY } = selectedAreaCoordinates;
+  if (direction === 'MoveUp') {
+    clientArea.scrollTop = topY - 10;
+  } else if (direction === 'MoveLeft') {
+    clientArea.scrollLeft = leftX - 10;
+  } else if (direction === 'MoveRight') {
+    clientArea.scrollLeft = leftX - clientArea.clientWidth / 2;
+  } else if (direction === 'MoveDown') {
+    clientArea.scrollTop = topY - clientArea.clientHeight / 2;
+  }
```
Comment (NatalliaHutovich) [inline packages/ketcher-react/src/script/editor/utils/canvasExtension.ts:167]: When structure is near the edge of screen and is moved out of visible area,  clientArea scroll may not keep up with the movement of the structure, so when mouse is released, we set scroll to make structure visible.  ToDo: Use clientArea.scrollTo with behavior: 'smooth' here.

**Comment by NatalliaHutovich** [comment]
ToDo: 1) Check scrolling and all calculation depending on zoom.
2) When there is an empty space between structure and edge of canvas, canvas is resizing to delete empty space. Find where this feature is called and delete (if need).
3) Make side panels disabled when autoscroll is started and mouse is on the panels.
4) Optional: use custom mouse pointer when selection is moving

**Comment by Nitvex** [comment]
The issue is fixed in scope of https://github.com/epam/ketcher/pull/3342

**Pull request closed without merging**